### PR TITLE
Make py_reader easier to use

### PR DESCRIPTION
**Pull request by sneaxiy**
This PR contains the following works:
- Modify `create_py_reader_by_data` API and add `prepend` parameter (default False). Users can set `prepend` be True, so that the created reader would be prepended to the program. This is pretty useful when the program is converted from other programs with pruning read_ops, or loaded from files.
- Add `remove_inference_program_feed_fetch_ops` to remove feed_op and fetch_op in inference models saved by `save_inference_model` API. This method returns the feed variables and fetch variables in the removed ops. After removing the feed_op and fetch_op, users can use feed or py_reader to test the program easily.
 
Traditionally, there are some inconvenience when using `py_reader`:
- Users have to use `program_guard` and `unique_name_guard` to create `train_program`, `train_reader` and `test_program`, `test_reader`.
- Users have to make two versions of codes to support `feed` and `py_reader`.
- Users have to give different names to different `py_reader`.

Now, user can use the following codes to create `py_reader`:
```
image = fluid.layers.data(name='image', ...)
label = fluid.layers.data(name='label', ...)
loss = network(image, label) # network definition when training
# convert train program to test program, no need to call unique_name_guard
test_main_prog = fluid.default_main_program().clone(for_test=True)
test_startup_prog = fluid.Program()

# prepend py_reader in front of program
train_reader = fluid.layers.create_py_reader_by_data(feed_list=[image, label], capacity=..., prepend=True)
with fluid.program_guard(test_main_prog, test_startup_prog):
   # Fluid would create unique names for different py_reader layers automatically.
   # Users do not need to give names to py_reader layers explicitly
   test_reader = fluid.layers.create_py_reader_by_data(feed_list=[image, label], capacity=..., prepend=True)

optimizer = fluid.optimizer.SGD(...)
optimizer.minimize(loss)

exe = fluid.Executor(fluid.CUDAPlace(0))
exe.run(fluid.default_startup_program())
exe.run(test_startup_prog)

train_reader.decorate_paddle_reader(...)
test_reader.decorate_paddle_reader(...)

# create parallel_executor for train program (i.e. fluid.default_main_program()) and test program (i.e. test_main_prog), and run in the following codes...
...
fluid.io.save_inference_model(...)
...
infer_model = fluid.io.load_inference_model(...)
feed_vars, fetch_vars = fluid.io.remove_inference_program_feed_fetch_ops(infer_model)
# Now, we can use py_reader or feed method to test the inference model !
```

More examples are performed in unittest test_py_reader_train_and_test.py

### CODE REVIEW

**Review by reyoung — DISMISSED** [review]

**Review by jacquesqiao — DISMISSED** [review]
LGTM!

File: paddle/fluid/API.spec [hunk 1]
```diff
@@ -5,6 +5,7 @@ paddle.fluid.Program.current_block ArgSpec(args=['self'], varargs=None, keywords
 paddle.fluid.Program.global_block ArgSpec(args=['self'], varargs=None, keywords=None, defaults=None)
 paddle.fluid.Program.list_vars ArgSpec(args=['self'], varargs=None, keywords=None, defaults=None)
 paddle.fluid.Program.parse_from_string ArgSpec(args=['binary_str'], varargs=None, keywords=None, defaults=None)
+paddle.fluid.Program.prune_read_op_and_convert_for_test ArgSpec(args=['self'], varargs=None, keywords=None, defaults=None)
```
Comment (panyx0718) [inline paddle/fluid/API.spec]: It can be a start of a suite of prune_xxx_op...
prune_send_op
prune_lr_op
prune_op_I_dont_like...

It will be nightmare

Comment (sneaxiy) [reply paddle/fluid/API.spec]: Remove it. Done.

**Review by reyoung — APPROVED** [review]

File: paddle/fluid/API.spec [hunk 2]
```diff
@@ -38,6 +38,7 @@ paddle.fluid.io.save_persistables ArgSpec(args=['executor', 'dirname', 'main_pro
 paddle.fluid.io.load_vars ArgSpec(args=['executor', 'dirname', 'main_program', 'vars', 'predicate', 'filename'], varargs=None, keywords=None, defaults=(None, None, None, None))
 paddle.fluid.io.load_params ArgSpec(args=['executor', 'dirname', 'main_program', 'filename'], varargs=None, keywords=None, defaults=(None, None))
 paddle.fluid.io.load_persistables ArgSpec(args=['executor', 'dirname', 'main_program', 'filename'], varargs=None, keywords=None, defaults=(None, None))
+paddle.fluid.io.remove_inference_program_feed_fetch_ops ArgSpec(args=['inference_program'], varargs=None, keywords=None, defaults=None)
```
Comment (panyx0718) [inline paddle/fluid/API.spec:41]: also remove this?

File: python/paddle/fluid/layers/io.py [hunk 3]
```diff
@@ -780,6 +792,7 @@ def create_py_reader_by_data(capacity,
        name(basestring): The prefix Python queue name and Reader name. None will
             be generated automatically.
        use_double_buffer(bool): Whether use double buffer or not.
+       prepend(bool): Whether to prepend the reader in front of the main program.
```
Comment (panyx0718) [inline python/paddle/fluid/layers/io.py:795]: when should this be true and when should be false?

Comment (sneaxiy) [reply python/paddle/fluid/layers/io.py:795]: When user wants to prepend `py_reader` in front of the program, it should be `True`, such as programs loaded from the previously saved ones without any read_ops.

**Pull request closed without merging**